llvm/lib/Target/AMDGPU/SIPreEmitPeephole.cpp

```diff
@@ -417,6 +454,293 @@ bool SIPreEmitPeephole::removeExeczBranch(MachineInstr &MI,
+    NewMI.addImm(SrcMO.getImm());
+    return;
+  }
+  // If op_sel == 0, select register 0 of reg:sub0_sub1
+  Register UnpackedSrcReg = (SrcMods & OpSelModifier)
+                                ? TRI->getSubReg(SrcMO.getReg(), AMDGPU::sub1)
+                                : TRI->getSubReg(SrcMO.getReg(), AMDGPU::sub0);
+
+  MachineOperand UnpackedSrcMO =
+      MachineOperand::CreateReg(UnpackedSrcReg, /*isDef=*/false);
+  if (SrcMO.isKill()) {
+    // For each unpacked instruction, mark its source registers as killed if the
+    // corresponding source register in the original packed instruction was
+    // marked as killed.
+    //
+    // Exception:
+    // If the op_sel and op_sel_hi modifiers require both unpacked instructions
+    // to use the same register (e.g., due to overlapping access to low/high
+    // bits of the same packed register), then only the *second* (latter)
+    // instruction should mark the register as killed. This is because the
+    // second instruction handles the higher bits and is effectively the last
+    // user of the full register pair.
+
+    bool OpSel = SrcMods & SISrcMods::OP_SEL_0;
+    bool OpSelHi = SrcMods & SISrcMods::OP_SEL_1;
+    bool KillState = true;
+    if ((OpSel == OpSelHi) && !IsHiBits)
+      KillState = false;
+    UnpackedSrcMO.setIsKill(KillState);
+  }
+  NewMI.add(UnpackedSrcMO);
+}
+
+void SIPreEmitPeephole::collectUnpackingCandidates(
+    MachineInstr &BeginMI, SetVector<MachineInstr *> &InstrsToUnpack,
+    uint16_t NumMFMACycles) {
+  auto *BB = BeginMI.getParent();
+  auto E = BB->end();
+  int TotalCyclesBetweenCandidates = 0;
+  auto SchedModel = TII->getSchedModel();
+  SetVector<Register> RegsOverlappedByMFMADef;
+  RegsOverlappedByMFMADef.insert(BeginMI.getOperand(0).getReg());
+  for (auto I = std::next(BeginMI.getIterator()); I != E; ++I) {
+    MachineInstr &Instr = *I;
+    if (Instr.isMetaInstruction())
+      continue;
+    if (Instr.isTerminator())
+      return;
+    if (TII->isNeverCoissue(Instr) && !isUnpackingSupportedInstr(Instr))
+      return;
+    const MCSchedClassDesc *InstrSchedClassDesc =
+        SchedModel.resolveSchedClass(&Instr);
+    TotalCyclesBetweenCandidates +=
+        SchedModel.getWriteProcResBegin(InstrSchedClassDesc)->ReleaseAtCycle;
+
+    if (TotalCyclesBetweenCandidates > NumMFMACycles)
+      return;
+    // Identify register dependencies between those used by the MFMA
+    // instruction and the following packed instructions. Also checks for
+    // transitive dependencies between the MFMA def and candidate instruction
+    // def and uses. Conservatively ensures that we do not incorrectly
+    // read/write registers.
+    for (const MachineOperand &InstrMO : Instr.operands()) {
+      if (InstrMO.isReg()) {
+        for (unsigned i = 0; i < RegsOverlappedByMFMADef.size(); ++i) {
+          if (TRI->regsOverlap(RegsOverlappedByMFMADef[i], InstrMO.getReg())) {
+            if (isUnpackingSupportedInstr(Instr))
```

Comment:
If there is any RAW or WAW dependency, we should `return` -- not just for the unpack supported instrs. We should be able to delete `RegsOverlappedByMFMADef`